pyqtorch/gpsr.py

```diff
@@ -161,9 +158,61 @@ def single_gap_shift(
                 / (4 * torch.sin(spectral_gap * shift / 2))
             )
 
-        def multi_gap_shift(*args, **kwargs) -> Tensor:
-            """Implements multi gap PSR rule."""
-            raise NotImplementedError("Multi-gap is not yet supported.")
+        def multi_gap_shift(
+            param_name: str,
+            values: dict[str, Tensor],
+            spectral_gaps: Tensor,
+            shift_prefac: Tensor = torch.tensor(0.5),
+        ) -> Tensor:
+            """Implements multi gap PSR rule.
+
+            Args:
+                param_name: Name of the parameter to apply PSR.
+                values: Dictionary with parameter values.
+                spectral_gaps: Spectral gaps value for PSR.
+                shift_prefac: Shift prefactor value for PSR shifts.
+                Defaults to torch.tensor(0.5).
+
+            Returns:
+                Gradient evaluation for param_name.
+            """
+            n_eqs = len(spectral_gaps)
+            PI = torch.tensor(torch.pi)
+            shifts = shift_prefac * torch.linspace(
+                PI / 2 - PI / 5, PI / 2 + PI / 5, n_eqs
+            )
+
+            # calculate F vector and M matrix
+            # (see: https://arxiv.org/pdf/2108.01218.pdf on p. 4 for definitions)
+            F = []
+            M = torch.empty((n_eqs, n_eqs))
+            n_obs = 1
+            for i in range(n_eqs):
+                shifted_values = values.copy()
+                shifted_values[param_name] = shifted_values[param_name] + shifts[i]
+                f_plus = expectation_fn(shifted_values)
+                shifted_values[param_name] = shifted_values[param_name] - 2 * shifts[i]
+                f_min = expectation_fn(shifted_values)
```

Comment:
```suggestion
                f_minus = expectation_fn(shifted_values)
```

Otherwise it could get confused with mininum and I guess we can afford two more letters.